HaloMD/master_server/src/main.rs

```diff
@@ -0,0 +1,205 @@
+    };
+
+    // Mutex for thread safety.
+    let servers_halo: Vec<HaloServer> = Vec::new();
+    let servers_mut_udp = Arc::new(Mutex::new(servers_halo));
+    let servers_mut_tcp = servers_mut_udp.clone();
+    let servers_mut_destruction = servers_mut_udp.clone();
+
+    // Destruction thread. This will remove servers that have not broadcasted their presence in a while.
+    thread::spawn(move || {
+        loop {
+            thread::sleep_ms(10 * 1000);
+            let mut servers = servers_mut_destruction.lock().unwrap();
+            let timenow = time::now().to_timespec().sec;
+            servers.retain(|x| x.last_alive + DROP_TIME > timenow);
+        }
+    });
+
+    // Blacklist mutex. Concurrency needs to be safe, my friend.
+    let blacklist: Vec<String> = Vec::new();
+    let blacklist_update = Arc::new(Mutex::new(blacklist));
+    let blacklist_udp = blacklist_update.clone();
+
+    // Blacklist read thread.
+    thread::spawn(move || {
+        loop {
+            // Placed in a block so blacklist is unlocked before sleeping to prevent threads from being locked for too long.
+            {
+                let mut blacklist_ref = blacklist_update.lock().unwrap();
+                match File::open(BLACKLIST_FILE) {
+                    Ok(file) => {
+                        let reader = BufReader::new(&file);
+                        match reader.lines().collect() {
+                            Err(_) => {
+                                blacklist_ref.clear();
+                            },
+                            Ok(t) => {
+                                *blacklist_ref = t;
+                            }
+                        }
+                    },
+                    Err(_) => {}
+                };
+            }
+            thread::sleep_ms(BLACKLIST_UPDATE_TIME * 1000);
+        }
+    });
+
+    // TCP server thread. This is for the HaloMD application.
+    thread::spawn(move || {
+        loop {
+            for stream in client_socket.incoming() {
+                let mut client = match stream {
+                    Err(_) => continue,
+                    Ok(the_stream) => the_stream
+                };
+                // Unwrap the IP.
+                let ip = match client.peer_addr() {
+                    Err(_) => continue,
+                    Ok(ip) => ip.to_string().split(":").next().unwrap().to_string()
+                };
+
+                let mut ips = String::new();
+
+                let servers_ref = servers_mut_tcp.lock().unwrap();
+                let servers = (*servers_ref).iter();
+
+                for j in servers {
+                    ips = ips + &(j.to_string()) + "\n";
+                }
+
+                // Some number placed after the requester's IP. If you ask me, the source code was abducted by aliens, and this is a tracking number. Regardless, it's needed.
+                ips = ips + &ip + ":49149:3425";
+                let _ = client.write_all(ips.as_bytes());
```

Comment:
`write_all` may block. Might consider spawning a thread for writing the all the data. Although, our current master server doesn't handle this, and we currently have a small server list. Still, this may be a good idea.